코드베이스로 탭바를 생성했습니다.

EuniceNam opened the pull request:
## 관련 이슈
#3 홈화면 UI를 구현

## 배경
루트뷰가 될 탭바를 만들었습니다. 

## 작업 내용
- [x] 검색 외 다른 탭 뷰컨트롤러 생성
- [x] 탭바컨트롤러 생성 및 커스텀
- [x] 시작 탭을 검색탭으로 설정
- [x] 탭바 상단 구분선 관련 설정
  - 스크롤뷰 위에서 기본적으로 구분선이 생겼다가, 맨 아래로 내려가면 구분선과 배경이 사라지는 것이 기본 옵션인 것 같습니다. 보더라인을 따로 설정했던 코드를 삭제하고, standardAppearance와 scrollEdgeAppearance에 배경색을 설정했습니다.   

### 스크린샷
![Simulator Screen Recording - iPhone 14 Pro - 2022-10-04 at 09 27 23](https://user-images.githubusercontent.com/18394923/193709215-c6fe05bf-1755-4dc5-8f0d-451962e61eda.gif)

<!--
이하는 꼭 채울 내용은 아님. 제목도 유동적으로 쓰기.
-->
## 리뷰 포인트
- 탭바를 이렇게 생성하는 것이 맞는지 다른분들 저장소에 배우러 가겠습니다. 
- [문화센터 비트와 리아의 커스텀 탭바 코드리뷰](https://github.com/DeveloperAcademy-POSTECH/CCC-1st-Toss-Bit/pull/6/files#r915613660)에서 뷰의 라이프사이클을 고려해서 탭바를 만든다는데 해당 코드는 이해를 못 해서 적용하지 않았습니다. 

## 참고
- [iOS15 탭바 보더라인 관련 글](https://madilyn.tistory.com/23)
- [다크모드 색 일부 안내](https://sarunw.com/posts/dark-color-cheat-sheet/)
- [노엘의 스눕클래스 이케아 클론코딩 탭바 PR](https://github.com/SnoopClass/Noel-Student-IKEA/pull/2/files#) - 축약형 인수 이름 ($0 ; shorthand argument names) 사용
- [브라운의 문화센터 토스 클론코딩의 탭바 PR](https://github.com/DeveloperAcademy-POSTECH/CCC-1st-Toss-Brown/pull/4/files#r912322348) - lazy var을 쓰는 이유
- [리아의 문화센터 토스 클론코딩의 탭바 PR](https://github.com/DeveloperAcademy-POSTECH/CCC-1st-Toss-Lia/pull/3/files#diff-0ad0c2125c8c53c8150df7b948be0f796a5578c7183edde83198597ef49e7959)
- [비트의 문화센터 토스 클론코딩의 탭바 PR](https://github.com/DeveloperAcademy-POSTECH/CCC-1st-Toss-Bit/pull/6/files#r915613660) - 뷰의 라이프사이클을 고려한 탭바

EuniceNam committed:
[ADD] 탭바 생성 (#3)

코드베이스로 탭바를 만들고, 그에 맞게 뷰컨트롤러 이름을 변경했습니다.
스토리보드 변경사항은 첫 화면에 연결된 클래스를 탭바컨트롤러로 바꾼 것뿐입니다.

EuniceNam committed:
[FIX] 탭바에서 구현하지 않을 탭뷰를 수정

구현하지 않을 탭뷰에 들어간 NavigationController를 지우고, 축약형 인수(shorthand argument names)를 사용해 모양새를 바꿨습니다.

Somin-DS reviewed (approved):
탭바 구현할때 lazy를 쓰는 이유를 몰랐었는데
첨부해주신 브라운의 코드리뷰를 보고 배웠습니다! 좀더 공부해봐야겠네요.
enum으로 탭바 아이콘 이름을 넣은 것도 좋은 것 같습니다!

EuniceNam committed:
[Fix] 탭바 보더 삭제를 통한 상단 구분선 설정

스크롤뷰 위에서 탭바의 상단 구분선이 기본으로 들어가므로, 보더 설정을 삭제하고, standardAppearance와 scrollEdgeAppearance에 배경색을 넣음.

unuhqueen commented on SnoopClass-IKEA/SearchViewController.swift:
```diff
@@ -7,12 +7,13 @@
 
     override func viewDidLoad() {
         super.viewDidLoad()
+        view.backgroundColor = .systemBackground
```
다크모드도 고려하셨군요! 👍

unuhqueen commented on SnoopClass-IKEA/TabBarController.swift:
```diff
@@ -0,0 +1,77 @@
+import UIKit
+
+// 아이콘의 SF symbol 이름 스트링
+enum TabIcon: String {
+    case home = "house"
+    case search = "rectangle.and.text.magnifyingglass"
+    case user = "person"
+    case wish = "heart"
+    case cart = "bag"
+}
```
sf symbol을 enum으로 만든게 넘 신박하네요!

compuTasha replied:
오 탭아이콘을 이렇게 관리하시는 이유가 있나요??

unuhqueen commented on SnoopClass-IKEA/TabBarController.swift:
```diff
@@ -0,0 +1,77 @@
+extension TabBarController {
+    func configureTabBar() {
+        // 탭바 모양 세팅
+        // 참고: https://madilyn.tistory.com/23
```
참고한 사이트를 주석으로 적어놓는 습관 좋은 것 같아요~

unuhqueen replied:
내용이 너무 충격입니다... 왜냐하면 저는 border가 나중에 사라지는 것도 몰랐기 때문... 유니스의 예리한 관찰력에 따봉을 드립니다....👍

EuniceNam replied:
저는 전에 들었는데 까먹고 있던 내용이라서요ㅎ 이브는 잊지 않으실 거라 믿어요!

compuTasha replied:
이 부분은 extension으로 뺀 이유가 있을까요?? 궁금해요!!

unuhqueen commented on SnoopClass-IKEA/TabBarController.swift:
```diff
@@ -0,0 +1,77 @@
+        let appearance = UITabBarAppearance()
+        appearance.backgroundColor = .systemBackground
+        appearance.selectionIndicatorTintColor = .label
+        tabBar.scrollEdgeAppearance = appearance
```
q.

제가 못 찾은 것 같긴 한데 위의 참고 사이트에서는 `let tabBar = UITabBar()` 선언 과정이 따로 있는데 여기서는 따로 없이 어떻게 가능한건가요?

EuniceNam replied:
- 해당 클래스가 UITabBarController를 상속해서 self.tabBar가 선언하지 않아도 있어서 가능합니다. 
- 참고 사이트에서 왜 선언했는지는 모르겠어요ㅠ 
  - 탭바를 만든 것도 같은데, if문 나가면 scope 문제로 없어질 것 같아서요. 
  - 참고로, UITabBar.appearance()는 전역변수(표현이 맞는지는 모르겠어요)라서 모든 탭바에 영향을 미칠 거에요.

unuhqueen commented on SnoopClass-IKEA/TabBarController.swift:
```diff
@@ -0,0 +1,77 @@
+class TabBarController: UITabBarController {
+    // 탭바의 각 탭
+    private lazy var homeViewController: UIViewController = {
+        $0.view.backgroundColor = .systemBrown
+        $0.tabBarItem = UITabBarItem(title: nil, image: UIImage(systemName: TabIcon.home.rawValue), tag: 0)
+        return $0
```
q.

여기서 어떻게 `$0`을 써도 코드가 돌아가는건지 클로저를 다시 공부해도 이해가 안가네요ㅠㅠㅠ 코멘트로 남기기 힘드시면 나중에 대면으로라도 알려주신다면 정말 감사하겠슴다..

EuniceNam replied:
Shorthand argument names (축약형|짧은 인수 이름) 으로 검색하시면 더 나올 수도 있어요. 
`{let vc = UIViewController() vc.어쩌구 ...}` 대신 `{$0.어쩌구}(UIViewController())`로 들어간 형태입니다.
스위프트가 클로저 인자에 $0, $1 등의 인수 이름을 자동으로 만드는데, 뒤의 UIViewController()가 클로저의 인자로 들어가기 때문이라고 합니다.
- $0...은 map등 고차함수에서 많이 쓴다고 합니다
- [a, b, c].forEach { $0.어쩌구 = 저쩌구 } 로 abc를 한 번에 돌아가게 쓰는 것도 재밌어요. 

출처
- [관련 사용법 안내(영어)](https://cocoacasts.com/shorthand-argument-names-in-swift)
- [공식문서 번역(에서 shorthand argument names 검색하기)](https://bbiguduk.gitbook.io/swift/language-guide-1/closures)
  - "Swift는 인라인 클로저에 $0, $1, $2 등 클로저의 인자값으로 참조하는데 사용할 수 있는 자동적으로 짧은 인자 이름 (shorthand argument names)을 제공합니다."

unuhqueen reviewed (commented):
좋은 코드 잘 봤습니다!! 덕분에 저도 많이 공부할 수 있었어요~~ 수고하셨습니다!

compuTasha reviewed (commented):
이제서야 확인했네요,,😓 고생하셨습니다!
코드 너무 깔끔해요!!

EuniceNam commented:
리뷰 감사해요! 
- 탭 아이콘은 SF심볼 이름보다 이넘 이름이 이해하기 편하고, 다른 아이콘으로 바꾸기 쉬워서 이넘을 써봤어요. 
  - 코드 동작에 필요한 스트링은 오타 때문에 주로 이넘으로 빼려고 하는데, 여기서는 재사용이 없어서 이점이 크지는 않은 것 같아요. 
  - 린트가 검색 아이콘 이름이 길다고 화를 낸 것도 있습니다ㅎ
- configureTabBar()의 상세내용은 class의 동작을 이해하는 데 필요하지 않을 것 같아서 extension에 넣었습니다. 
  - 저는 extension을 내용 구분을 위해서와 중요하지 않은 내용을 숨기기 위해서 사용하고 있어요. 
    - 오히려 복잡하게 하는 것 같다면 알려주세요:)
    - extension의 원래 용도는 더 있을 것 같아요. 
    - delegate 등으로 내용이 구분되는 extension이 따로 있었다면 configureTabBar()는 분리하지 않았을 수 있을 것 같아요.

compuTasha commented:
> 리뷰 감사해요!
> 
> * 탭 아이콘은 SF심볼 이름보다 이넘 이름이 이해하기 편하고, 다른 아이콘으로 바꾸기 쉬워서 이넘을 써봤어요.
>   
>   * 코드 동작에 필요한 스트링은 오타 때문에 주로 이넘으로 빼려고 하는데, 여기서는 재사용이 없어서 이점이 크지는 않은 것 같아요.
>   * 린트가 검색 아이콘 이름이 길다고 화를 낸 것도 있습니다ㅎ
> * configureTabBar()의 상세내용은 class의 동작을 이해하는 데 필요하지 않을 것 같아서 extension에 넣었습니다.
>   
>   * 저는 extension을 내용 구분을 위해서와 중요하지 않은 내용을 숨기기 위해서 사용하고 있어요.
>     
>     * 오히려 복잡하게 하는 것 같다면 알려주세요:)
>     * extension의 원래 용도는 더 있을 것 같아요.
>     * delegate 등으로 내용이 구분되는 extension이 따로 있었다면 configureTabBar()는 분리하지 않았을 수 있을 것 같아요.

답변 감사합니다!
extension의 용도에 대해 정리한 글이 있어 첨부합니다!
https://babbab2.tistory.com/124
https://blog.yagom.net/529/
